packages/plugins-image/src/components/Image.js

```diff
@@ -0,0 +1,206 @@
+import React from "react";
+import propTypes from "prop-types";
+import DefaultImageRenderer from "./DefaultImageRenderer";
+import validImageFormats from "../static/validImageFormats";
+
+export const bytesToMb = bytes => `${(bytes / 1048576).toFixed(1)} MB`;
+
+/* eslint no-param-reassign: ["error", { "props": false }] */
+export const resetForm = input => {
+  if (input) {
+    input.value = "";
+  }
+};
+
+const makeCancelable = promise => {
+  let hasCanceled = false;
+
+  const wrappedPromise = new Promise((resolve, reject) => {
+    promise.then(
+      val => (hasCanceled ? reject(new Error("Cancelled")) : resolve(val)),
+      error => (hasCanceled ? reject(new Error("Cancelled")) : reject(error))
+    );
+  });
+
+  return {
+    promise: wrappedPromise,
+    cancel() {
+      hasCanceled = true;
+    }
+  };
+};
+
+class Image extends React.Component {
+  constructor(props) {
+    super(props);
+    const src = props.node.data.get("src") || "";
+    this.state = { src, loading: true, error: "" };
+    this.attemptBlobUpload();
+  }
+  componentWillUnmount() {
+    if (this.uploader) this.uploader.cancel();
+  }
+
+  onImgLoad = () => {
+    // Non-blobs onLoad should finish loading
+    if (!this.state.src.startsWith("blob:")) {
+      this.setState({ loading: false });
+    }
+  };
+
+  attemptBlobUpload() {
+    const { src } = this.state;
+    if (src && src.includes("blob")) {
+      fetch(src)
+        .then(resp => resp.blob())
+        .then(blob => {
+          const file = new File([blob], `clipboard.png`);
+          this.attemptUpload(file);
+        });
+    }
+  }
+
+  attemptUpload(file) {
+    this.setState({ loading: true });
+    const { uploadImage } = this.props.options;
+    if (uploadImage) {
+      this.uploader = makeCancelable(uploadImage(file));
+      this.uploader.promise
+        .then(newUrl => {
+          this.updateSrc(newUrl);
+        })
+        .catch(() => {
+          this.updateSrc("");
+          this.updateError(`Failed to upload file to server`);
```

Comment:
This triggers two `setState` in quick successions. Could combine updateError and updateSrc